static/scripts/rewards/web3/connect-wallet.ts

```diff
@@ -18,18 +22,148 @@ export async function connectWallet(): Promise<JsonRpcSigner | null> {
       return null;
     }
 
+    const isOkay = await stressTestWalletRpc(wallet);
+
+    if (!isOkay) {
+      if (mediaQuery.matches) {
+        toaster.create("info", `In case of network issues, please change your in-wallet RPC to the one below...`, 15000);
+      } else {
+        // Their wallet provider will auto-prompt due to the call succeeding
+        toaster.create("error", "We have detected potential issues with your in-wallet RPC. Accept the request to replace it with a more reliable one.");
+      }
+      await addFastestHandlerNetwork(wallet);
+    }
+
     return signer;
   } catch (error: unknown) {
     return connectErrorHandler(error);
   }
 }
 
+async function addFastestHandlerNetwork(wallet: ethers.providers.Web3Provider) {
+  const networkId = app.networkId ?? (await wallet.getNetwork()).chainId;
+  const handler = useHandler(networkId);
+  let provider = await handler.getFastestRpcProvider();
+  const appUrl = app.provider?.connection?.url;
+
+  const latencies = handler.getLatencies();
+  const latenciesArray = Object.entries(latencies).map(([url, latency]) => ({ url, latency }) as { url: string; latency: number });
+  const sorted = latenciesArray.sort((a, b) => a.latency - b.latency);
+
+  let toSuggest = sorted[0];
+
+  let isOkay = false;
+
+  for await (const { url } of sorted) {
+    const _url = url.split("__")[1];
+    if (_url !== appUrl) {
+      provider = new ethers.providers.JsonRpcProvider(_url);
+
+      isOkay = await stressTestWalletRpc(provider);
+
+      if (isOkay) {
+        toSuggest = { url: _url, latency: latencies[url] };
+        break;
+      }
+    }
+  }
+
+  if (!isOkay) {
+    toaster.create("error", "We failed to find a more reliable RPC for you. Please try again later if you have network issues.");
+    return;
+  }
+
+  try {
+    await addHandlerSuggested(wallet, toSuggest.url);
+  } catch (error) {
+    toaster.create("info", `${toSuggest.url}`, Infinity);
+  }
+}
+
+async function addHandlerSuggested(provider: ethers.providers.Web3Provider, url: string) {
+  const symbol = app.networkId === 1 ? "ETH" : "XDAI";
+  const altSymbol = app.networkId === 1 ? "eth" : "xdai";
+  const altSymbol2 = app.networkId === 1 ? "Eth" : "xDai";
+
+  if (mediaQuery.matches) {
```

Comment:
You are assuming it is a mobile browser based on the media query. 

I'm sure there is a better way to do this in typescript.

Reply:
> While media queries are the most straightforward and widely supported method for responsive design, relying solely on them can be limiting. A combination approach, leveraging User-Agent detection and feature detection alongside media queries, can provide a more comprehensive and adaptable solution.

None that are as reliable as this, you are relying on detection of env specific globals or useragent checks and with the state of Web3 mobile I felt this was a flakier approach

Reply:
```typescript
  const isTouchScreen = 'ontouchstart' in window || navigator.maxTouchPoints > 0;
```

I'm sure there are npm modules that handle this check robustly. I don't think it makes sense to rely on the media query in a javascript runtime.